### ISSUE TITLE
All "resnet-v2-50" are same architecture?

### ISSUE THREAD

**Issue by kento-kobayashi** [tensorflow/hub#754]
Are Model architectures same resnet in https://tfhub.dev/s?network-architecture=resnet-v2-50 ?
Many biases are used in https://tfhub.dev/google/imagenet/resnet_v2_50/feature_vector/4 (and https://tfhub.dev/google/imagenet/resnet_v2_50/feature_vector/4) like `model_weights/keras_layer/resnet_v2_50/block1/unit_1/bottleneck_v2/conv3/biases:0`. But, https://tfhub.dev/google/remote_sensing/uc_merced-resnet50/1 and https://tfhub.dev/google/remote_sensing/uc_merced-resnet50/1 use only one bias like `model_weights/keras_layer/resnet/head/conv2d/bias`.
Why many biases are used in ImageNet-model and not used in other models?

**Comment by arghyaganguly** [comment]
TensorFlow Hub Module provides us with graph comprising of architecture of model along with it’s weights and biases trained on certain datasets.
Even though the underlying vanilla architecture is same for all resnet-v2-50, the number of bias being used will differ depending on the dataset the model was trained on based on the use-case.

Different datasets will have different attributes/featurespace(gradient magnitude,color,grayscale intensity,edges and others) which will be captured/stored in different feature vectors.

**Comment by kento-kobayashi** [comment]
Thank you very much for your quick reply.
Is https://github.com/tensorflow/models/blob/master/research/slim/nets/resnet_v2.py vanilla architecture? 
I checked each models bellow script. What is the vanilla?

```
import tensorflow as tf
import tensorflow_hub as hub
import h5py

# model_name = "remote_sensing_uc_merced-resnet50_1"
# model_name = "remote_sensing_eurosat-resnet50_1"
# model_name = "imagenet_resnet_v2_50_feature_vector_1"
model_name = "imagenet_resnet_v2_50_feature_vector_4"

model = hub.KerasLayer(f"{model_name}", dynamic=True)
model = tf.keras.Sequential([model])
model.save(f"{model_name}_debug/{model_name}.h5")


f = h5py.File(f"{model_name}_debug/{model_name}.h5", "r")


def check_value(x):
    if hasattr(f[x], "value") and "bias" in x:
        print(x, f[x].value.shape)


f.visit(check_value)
```

For `remote_sensing_uc_merced-resnet50_1`
```
model_weights/keras_layer/resnet/head/conv2d/bias:0 (21,)
```

For `remote_sensing_eurosat-resnet50_1`
```
model_weights/keras_layer/resnet/head/conv2d/bias:0 (10,)
```

For `imagenet_resnet_v2_50_feature_vector_1`
```
model_weights/keras_layer/resnet_v2_50/block1/unit_1/bottleneck_v2/conv3/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block1/unit_1/bottleneck_v2/shortcut/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block1/unit_2/bottleneck_v2/conv3/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block1/unit_3/bottleneck_v2/conv3/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block2/unit_1/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_1/bottleneck_v2/shortcut/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_2/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_3/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_4/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block3/unit_1/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_1/bottleneck_v2/shortcut/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_2/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_3/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_4/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_5/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_6/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block4/unit_1/bottleneck_v2/conv3/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/block4/unit_1/bottleneck_v2/shortcut/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/block4/unit_2/bottleneck_v2/conv3/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/block4/unit_3/bottleneck_v2/conv3/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/conv1/biases:0 (64,)
```

For `imagenet_resnet_v2_50_feature_vector_4`
```
model_weights/keras_layer/resnet_v2_50/block1/unit_1/bottleneck_v2/conv3/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block1/unit_1/bottleneck_v2/shortcut/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block1/unit_2/bottleneck_v2/conv3/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block1/unit_3/bottleneck_v2/conv3/biases:0 (256,)
model_weights/keras_layer/resnet_v2_50/block2/unit_1/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_1/bottleneck_v2/shortcut/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_2/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_3/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block2/unit_4/bottleneck_v2/conv3/biases:0 (512,)
model_weights/keras_layer/resnet_v2_50/block3/unit_1/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_1/bottleneck_v2/shortcut/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_2/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_3/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_4/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_5/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block3/unit_6/bottleneck_v2/conv3/biases:0 (1024,)
model_weights/keras_layer/resnet_v2_50/block4/unit_1/bottleneck_v2/conv3/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/block4/unit_1/bottleneck_v2/shortcut/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/block4/unit_2/bottleneck_v2/conv3/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/block4/unit_3/bottleneck_v2/conv3/biases:0 (2048,)
model_weights/keras_layer/resnet_v2_50/conv1/biases:0 (64,)
```

**Comment by arghyaganguly** [comment]
@kento-kobayashi ,please find the below information regarding Resnet architechture :- 

1) dataset :- Imagenet
a)https://tfhub.dev/google/imagenet/resnet_v2_50/feature_vector/1 :- hub module (v1) usage
b)https://tfhub.dev/google/imagenet/resnet_v2_50/feature_vector/4 :- hub module (v2) ​usage

2) dataset :- UC Merced remote sensing dataset
https://tfhub.dev/google/remote_sensing/uc_merced-resnet50/1 :- hub module(v1) usage

3) dataset :- EuroSat remote sensing dataset 
https://tfhub.dev/google/remote_sensing/eurosat-resnet50/1 :- hub module(v1) usage


Resnet architechtures :- 

1) resnet_50 architechture(v1) :- [link](https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet50.py)

2) resnetv2 ( Resnet trained on Imagenet dataset can be used for transfer learning) -> [link](https://github.com/keras-team/keras-applications/blob/master/keras_applications/inception_resnet_v2.py)

3) resnetv2(Resnet plain vanilla not pre trained for Keras ) -> [link](https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet_v2.py) 

4) resnet_v2(Resnet pain vanilla not pre-trained for tf_slim (a different high level api than keras)) --> [link](https://github.com/tensorflow/models/blob/master/research/slim/nets/resnet_v2.py)

5) vanilla is not pre-trained on any dataset example COCO or Imagenet etc.
If you want the keras implementation it is available in   [resnet_common.py](https://github.com/keras-team/keras-applications/blob/master/keras_applications/resnet_common.py)
(Stack type 1 = Resnetv1, stack type 2 = Resnetv2, stack type 3 = ResNext)






note:The TF1-style API of TensorFlow Hub works with the v1 compatibility mode of TensorFlow 2.

**Comment by kento-kobayashi** [comment]
@arghyaganguly 
I tried to compare models which have exactly the same sets of parameters and are trained on different datasets. But I cannot do that with the currently available parameters. So I think it is confusing to say that those models have the same architecture even though they have different sets of parameters.

**Comment by MorganR** [comment]
Thanks for your question @kento-kobayashi! When labeling model architectures on TF Hub, we don't guarantee that all models with the same architecture are using 100% the same parameters, but instead aim to capture the overall structure.

As you pointed out, the uc_merced and eurosat models only include bias variables on the head layers, whereas the imagenet versions include them on all layers. However, these models are all still ResNet v2 50 architectures. You can find more specifics about each model by reading the reference material linked on each model's details page ([UC_merced and eurosat paper](https://arxiv.org/pdf/1911.06721.pdf) and [TF Slim model details](https://github.com/tensorflow/models/blob/master/research/slim/README.md#pre-trained-models)).

We mostly expect users of TF Hub to treat models as blackboxes with clearly defined inputs and outputs. If a model meets your needs (eg. accuracy, speed, size, pre-training category), then use it! If you instead want to dig into the internal details of a model's structure, then you may be more interested in code-first resources, such as Model Garden or Keras applications. These often come with pre-trained checkpoints that can be loaded with the provided code.

I'm not entirely sure what you mean by the "vanilla" implementation of Resnet V2 50. The [original ResNet V2 paper](https://arxiv.org/pdf/1603.05027.pdf) links to [this Lua code](https://github.com/KaimingHe/resnet-1k-layers). If you want a TF implementation, then any of the Model Garden or Keras links provided by @arghyaganguly could work for you.

Does that fully answer your question?

**Comment by arghyaganguly** [comment]
@kento-kobayashi , please confirm on the above comment by @MorganR.
Thanks.

**Comment by kento-kobayashi** [comment]
Thank you for everything @arghyaganguly and @MorganR.

> Does that fully answer your question?
Yes! I understood clearly.
I could good results from your answer and tfhub. I really appreciate all of you!

**Comment by arghyaganguly** [comment]
Closing this based on the confirmation from @kento-kobayashi. Thanks.
